### ISSUE TITLE
Map incoming header to more refined customer header

### ISSUE THREAD

**Issue by Mrc0113** [spring-cloud/spring-cloud-stream#2120]
It would be nice if the Cloud Stream framework would provide a way to map incoming headers to a more refined custom header. I'd imagine this would probably use a combination of SPEL and existing header enricher capabilities in Spring Integration? And it would be ideal if this could take place completely via configuration (if possible) to prevent the developer from having messaging specific code. 

This would enable developers to do many things, but one example would be to extract parts of a destination as separate header variables as requested in a Solace Binder enhancement here: https://github.com/SolaceProducts/solace-spring-cloud/pull/45. It would be great if this functionality could be achieved at the framework level so everyone can benefit from it. Note that this would also assist with folks discovering Spring Cloud Stream via AsyncAPI since it has the concept (like OpenAPI) of variables that can exist in channels. 

FYI @olegz this is one of the enhancements we discussed earlier.

**Comment by olegz** [comment]
This may already be supported via SI header enricher mechanism if we can wire to it (probably via channel interceptor)

**Comment by garyrussell** [comment]
This is already supported in the KafkaBinder via the `headerMapperBeanName` but something standard that would apply across all binders would be nice.

**Comment by olegz** [comment]
Starting to look at this. . . I believe this is a bit different (at least semantically) to HeaderMapper. It's more of a HeaderEnricher since the output of this would always be some kind of key/value structure (e.g., Map) that would augment/change the existing headers.

**Comment by Mrc0113** [comment]
@olegz, yes the proposed solution on the PR in the Solace repo currently stores the output as a Map, but I think there are too options:
1. Store the output in a map
2. Just store each as it's own defined header

Each option has pros and cons. The first option would likely be saved in a reserved header space and therefore you don't have to deal with conflicts. However that may not work as well with Function beans that are being used with Cloud Stream & non-Cloud Stream apps.  The second option would allow you to map/create new headers which enables probably better re-use of Function beans across different types of apps, but then we have to decide what to do in the case of a conflict. 

I think I would actually learn towards the latter...maybe it could even be used to put things in a map if the developer desired that approach?

**Comment by olegz** [comment]
Transferring it to s-c-function . We can actually already do this. There is a component in SCF that one can register as bean. But I would need to document it and do some polishing. . .
